docs/playground/README.md

```diff
@@ -0,0 +1,90 @@
+
+### Mockchain pane
+
+The mockchain pane has three major parts.
+
+1. A visual representation of the mockchain in the form of the UTxO graph. The UTxO graph consists of nodes that represent individual (valid) transactions and edges that represent (valid) input-output connections.
+2. The final assignment of funds to wallets. This may be a subsets of the overall funds, because some funds may be locked by scripts. Optionally, the wallets should be colour coded and edges paying funds to those wallets should use the same colour. Moreover, we may display the wallet funds as coloured bars whose relative length represents their relative (final) value.
+3. A possibly empty list of *invalid* transactions, each of which indicates the reason for why it is invalid.
+
+Optionally, provide introspection facilities for valid and invalid transactions and graph edges. In particular, to query information that helps with debugging and also includes rejected, invalid transactions and failed script execution.
+
+
+## Protocol between the frontend and backend
+
+We are using a stateless architecture. Hence, the entire frontend state (that affects contract execution) needs to be transmitted to the backend on each request. Central is the contract source code and the current sequence of *wallet events*. A wallet event comprises the name of a *wallet endpoint function* and a list of arguments matching the type of the wallet endpoint function. Wallet endpoint functions are those designated functions within a contract that can be invoked by the user of a wallet that has loaded the contract.
+
+### Frontend state required for contract execution
+
+1. The contents of all Plutus source modules. (Initially, this is one module; optionally, we will extend it to a list of modules subsequently.)
+2. List of initial amount of ADA in each wallet. (The length of this list determines the number of wallets.)
+3. List of wallet events.
+
+### Wallet events
+
+The wallet events are invocations of wallet endpoint functions (aka *contract endpoints*), where the wallet user provides arguments to the endpoint function. The user does so within the frontend UI, which requires a limit on the number and form of those arguments. We support the following types as arguments types for wallet endpoint function s:
+
+* `String`,
+* `PubKey`,
+* integral types, and
+* products of the above.
+
+Products can be in the form of record datatypes with a single constructor. Wallet endpoint functions can, at most, have TBD arguments.
+
+## Contracts
+
+Every contract includes a module `Contract` that exports a main `contract` function. The `contract` function uses 
+
+```
+registerEndpoint :: EndpointFunction fn => fn -> IO ()
```

Comment:
I'm worried about making this `IO`. It's a security hole. Can we constrain it more tightly?

I don't think saying, "It's just a throwaway small AWS instance," is enough. If someone can compromise a box, they can compromise anything that's subsequently sent to that box. (eg. Sniffing other people's contracts.)

I suppose the only material risk is repetitional damage, but that can't be overlooked...

Reply:
Yes, you are right. The actual coordination code runs in its own monad. We want something different here, but something more specialised than `IO`. I am not quite sure what yet, but I'll put in a fictional placeholder type to clarify that we don't want full `IO`.